File: web/scripts/seed-qdrant.mjs
/**
 * Seeds Qdrant Cloud with the 30 historical fraud cases from fixtures.
 * Run once after creating your Qdrant cluster:
 *
 *   QDRANT_URL=https://xxx.qdrant.io QDRANT_API_KEY=your_key node scripts/seed-qdrant.mjs
 */
import { QdrantClient } from "@qdrant/js-client-rest";
import { createHash } from "crypto";
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dir = dirname(fileURLToPath(import.meta.url));
const cases = JSON.parse(readFileSync(join(__dir, "../fixtures/historical_fraud_cases.json"), "utf8"));

const QDRANT_URL = process.env.QDRANT_URL;
const QDRANT_API_KEY = process.env.QDRANT_API_KEY;
const COLLECTION = process.env.QDRANT_COLLECTION || "fraud_case_memory";
const DIM = 128;

if (!QDRANT_URL || !QDRANT_API_KEY) {
  console.error("Set QDRANT_URL and QDRANT_API_KEY env vars.");
  process.exit(1);
}

function embed(text) {
  const vec = new Array(DIM).fill(0);
  const words = text.toLowerCase().match(/[a-z]+/g) || [];
  for (const w of words) {
    const digest = createHash("md5").update(w).digest("hex");
    const idx = parseInt(digest.slice(0, 8), 16) % DIM;
    vec[idx] += 1;
  }
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
  return norm > 0 ? vec.map((v) => v / norm) : vec;
}

function amountBracket(amount) {
  if (amount < 1000) return "UNDER_1K";
  if (amount < 10000) return "1K_10K";
  if (amount < 50000) return "10K_50K";
  return "OVER_50K";
}

function caseIdToInt(id) {
  return Math.abs(id.split("").reduce((h, c) => (h * 31 + c.charCodeAt(0)) | 0, 0));
}

const client = new QdrantClient({ url: QDRANT_URL, apiKey: QDRANT_API_KEY });

// Create or recreate collection
try {
  await client.deleteCollection(COLLECTION);
  console.log("Deleted existing collection.");
} catch {}

await client.createCollection(COLLECTION, {
  vectors: { size: DIM, distance: "Cosine" },
});
console.log(`Created collection '${COLLECTION}' (${DIM}-dim, Cosine).`);

// Upsert all cases
const points = cases.map((c) => ({
  id: caseIdToInt(c.case_id),
  vector: embed(c.narrative),
  payload: {
    case_id: c.case_id,
    narrative: c.narrative,
    fraud_type: c.fraud_type,
    amount_bracket: amountBracket(c.amount || 0),
    channel: c.channel,
    geography: c.geography,
    resolution_date: c.resolution_date,
    analyst_verdict: c.analyst_verdict,
  },
}));

await client.upsert(COLLECTION, { wait: true, points });
console.log(`Seeded ${points.length} cases into Qdrant.`);

// Quick sanity search
const result = await client.search(COLLECTION, {
  vector: embed("offshore wire transfer high risk"),
  limit: 3,
  with_payload: true,
});
console.log("\nTop 3 results for 'offshore wire transfer high risk':");
result.forEach((r) => console.log(`  ${r.payload.case_id}  score=${r.score.toFixed(3)}  ${r.payload.fraud_type}`));
console.log("\nDone. Add QDRANT_URL and QDRANT_API_KEY to Vercel env vars.");

File: scripts/seed_qdrant.py
#!/usr/bin/env python3
"""
Seeds Qdrant with the historical fraud case memory (PRD section 6 schema:
fraud_type, amount_bracket, channel, geography, resolution_date), so the
Fraud Case Search Agent has real precedent cases to retrieve against.

Run from the repo root, against the docker-compose-exposed Qdrant port:
    pip install -r scripts/requirements.txt
    python scripts/seed_qdrant.py
"""
import hashlib
import json
import os
import re
import sys

import numpy as np
from qdrant_client import QdrantClient
from qdrant_client.models import Distance, PointStruct, VectorParams

QDRANT_URL = os.environ.get("QDRANT_URL", "http://localhost:6333")
COLLECTION = os.environ.get("QDRANT_COLLECTION", "fraud_case_memory")
DIM = 128

HERE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def embed(text: str):
    vec = np.zeros(DIM, dtype=np.float32)
    for word in re.findall(r"[a-zA-Z]+", text.lower()):
        idx = int(hashlib.md5(word.encode()).hexdigest(), 16) % DIM
        vec[idx] += 1.0
    norm = np.linalg.norm(vec)
    if norm > 0:
        vec = vec / norm
    return vec.tolist()


def amount_bracket(amount: float) -> str:
    if amount < 1000:
        return "UNDER_1K"
    if amount < 10000:
        return "1K_10K"
    if amount < 50000:
        return "10K_50K"
    return "OVER_50K"


def main():
    with open(os.path.join(HERE, "fixtures", "historical_fraud_cases.json")) as f:
        cases = json.load(f)

    client = QdrantClient(url=QDRANT_URL)
    collections = [c.name for c in client.get_collections().collections]
    if COLLECTION not in collections:
        client.create_collection(
            collection_name=COLLECTION,
            vectors_config=VectorParams(size=DIM, distance=Distance.COSINE),
        )

    points = []
    for i, case in enumerate(cases):
        vector = embed(case["narrative"])
        points.append(
            PointStruct(
                id=i + 1,
                vector=vector,
                payload={
                    "case_id": case["case_id"],
                    "narrative": case["narrative"],
                    "fraud_type": case["fraud_type"],
                    "amount_bracket": amount_bracket(case["amount"]),
                    "channel": case["channel"],
                    "geography": case["geography"],
                    "resolution_date": case["resolution_date"],
                    "analyst_verdict": case["analyst_verdict"],
                },
            )
        )

    client.upsert(collection_name=COLLECTION, points=points)
    print(f"Seeded {len(points)} historical fraud cases into Qdrant collection '{COLLECTION}'.")


if __name__ == "__main__":
    main()
